blockmanager.go

```diff
@@ -595,6 +595,96 @@ func (b *blockManager) handleDonePeerMsg(peers *list.List, sp *serverPeer) {
 	}
 }
 
+// errToWireRejectCode determines the wire rejection code and description for a
+// given error. This function can convert some select blockchain and mempool
+// error types to the historical rejection codes used on the p2p wire protocol.
+func errToWireRejectCode(err error) (wire.RejectCode, string) {
+	if err == nil {
+		panic("errToWireRejectCode(): err should not be nil")
+	}
+
+	// Unwrap mempool errors.
+	if rerr, ok := err.(mempool.RuleError); ok {
+		err = rerr.Err
+	}
+
+	// The default reason to reject a transaction/block is due to it being
+	// invalid somehow.
+	code := wire.RejectInvalid
+	var reason string
+
+	switch err := err.(type) {
+	case blockchain.RuleError:
+		// Convert the chain error to a reject code.
+		switch err.ErrorCode {
+		// Rejected due to duplicate.
+		case blockchain.ErrDuplicateBlock:
+			code = wire.RejectDuplicate
+
+		// Rejected due to obsolete version.
+		case blockchain.ErrBlockVersionTooOld:
+			code = wire.RejectObsolete
+
+		// Rejected due to checkpoint.
+		case blockchain.ErrCheckpointTimeTooOld:
```

Comment:
It doesn't matter much, but this could all be much more compact (and readable imo) using commas for the additional cases instead of `fallthrough`.

e.g.
```Go
	switch err.ErrorCode {
	case blockchain.ErrCheckpointTimeTooOld, blockchain.ErrDifficultyTooLow,
		blockchain.ErrBadCheckpoint, blockchain.ErrForkTooOld:

		code = wire.RejectCheckpoint
	}

	reason = err.Error()
```

The same holds true for the others.